src/vm/assembly.cpp

```diff
@@ -1780,6 +1780,37 @@ INT32 Assembly::ExecuteMainMethod(PTRARRAYREF *stringArgs, BOOL waitForOtherThre
     INT32   iRetVal = 0;
 
     BEGIN_ENTRYPOINT_THROWS;
+    {
+        GCX_PREEMP();
+
+        mdToken tkEntryPoint = m_pManifestFile->GetEntryPointToken();
+        // <TODO>@TODO: What if the entrypoint is in another file of the assembly?</TODO>
+        ReleaseHolder<IMDInternalImport> scope(m_pManifestFile->GetMDImportWithRef());
+        // In theory, we should have a valid executable image and scope should never be NULL, but we've been  
+        // getting Watson failures for AVs here due to ISVs modifying image headers and some new OS loader 
+        // checks (see Dev10# 718530 and Windows 7# 615596)
+        if (scope == NULL)
+        {
+            ThrowHR(COR_E_BADIMAGEFORMAT);
+        }
+
+#ifdef FEATURE_COMINTEROP 
+        Thread::ApartmentState state = Thread::AS_Unknown;        
+
+        if((!IsNilToken(tkEntryPoint)) && (TypeFromToken(tkEntryPoint) == mdtMethodDef)) {
```

Comment:
This should just use the `MethodDesc*` of the entrypoint that is fetched below:  pMeth = GetEntryPoint();. No need to compute it or validate the token twice.

Reply:
I have removed the checks. but I am not able to figure how to avoid computing tkentrypoint(mdtoken)
mdToken tkEntryPoint = m_pManifestFile->GetEntryPointToken();
As we are using this token to get the state in 
state = SystemDomain::GetEntryPointThreadAptState(scope, tkEntryPoint);